### PR TITLE
[PR-14360] Refactor TKBLContactsLoader, add topViewController snippet

### PR DESCRIPTION
**Pull request by kostilal**
## Related Stories
[](https://talkable.atlassian.net/browse/PR-14360)

### CODE REVIEW

File: Talkable/TKBLContactsLoader.h [hunk 1]
```diff
@@ -16,8 +16,6 @@ FOUNDATION_EXPORT NSString* TKBLContactPhoneNumberKey;
 
 @interface TKBLContactsLoader : NSObject
 
-+ (instancetype)loader;
-
-- (void)loadContactsWithComplitionHandler:(void(^)(NSArray* contactList))complitionHandler;
+- (void)loadContactsWithComplitionHandler:(void(^)(NSArray* contacts))complitionHandler;
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.h]:
```suggestion
- (void)loadContactsWithCompletionHandler:(void(^)(NSArray* contacts))completionHandler;
```

File: Talkable/TKBLContactsLoader.m [hunk 2]
```diff
@@ -20,87 +21,110 @@
...
-        TKBLLog(@"error while loading contacts - %@", CFBridgingRelease(error));
-        return;
-    }
-
-    ABAddressBookRequestAccessWithCompletion(addressBook, ^(bool granted, CFErrorRef error) {
-        if (error)
-            TKBLLog(@"error while loading contacts - %@", CFBridgingRelease(error));
-
+- (void)loadContactsWithComplitionHandler:(void(^)(NSArray* contacts))complitionHandler {
+    [self requestForAccessWithComplition:^(BOOL granted) {
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
- (void)loadContactsWithCompletionHandler:(void(^)(NSArray* contacts))completionHandler {
    [self requestForAccessWithCompletion:^(BOOL granted) {
```

File: Talkable/TKBLContactsLoader.m [hunk 3]
```diff
@@ -20,87 +21,110 @@
...
         if (granted) {
-            NSArray* contacts = [self grabContactsFromAddressBook:addressBook];
+            NSArray *contacts = [self grabContacts];
             dispatch_async(dispatch_get_main_queue(), ^{
                 complitionHandler(contacts);
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
                completionHandler(contacts);
```

File: Talkable/TKBLContactsLoader.m [hunk 4]
```diff
@@ -20,87 +21,110 @@
...
             });
         }
-        CFRelease(addressBook);
-    });
+    }];
 }
 
-#pragma mark - [Private]
+- (void)presentPermissionAlert {
+    NSString *message = NSLocalizedString(@"This app requires access to your contacts to function properly. Please visit to the Privacy section in the Settings app.", nil);
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
    NSString *message = NSLocalizedString(@"This app requires access to your contacts to function properly. Please visit the Privacy section in the Settings app.", nil);
```

File: Talkable/TKBLContactsLoader.m [hunk 5]
```diff
@@ -20,87 +21,110 @@
...
+}
 
-        [contacts addObject:@{
-            TKBLContactFirstNameKey: firstName ? firstName : [NSNull null],
-            TKBLContactLastNameKey: lastName ? lastName : [NSNull null],
-            TKBLContactFullNameKey: fullName ? fullName : [NSNull null],
-            TKBLContactEmailKey: emails,
-            TKBLContactPhoneNumberKey: phoneNumbers
-        }];
+- (void)requestForAccessWithComplition:(void(^)(BOOL granted))complitionHandler {
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
- (void)requestForAccessWithCompletion:(void(^)(BOOL granted))completionHandler {
```

File: Talkable/TKBLContactsLoader.m [hunk 6]
```diff
@@ -20,87 +21,110 @@
...
+    CNAuthorizationStatus status = [CNContactStore authorizationStatusForEntityType:CNEntityTypeContacts];
+    CNContactStore *store = [[CNContactStore alloc] init];
+    
+    switch (status) {
+        case CNAuthorizationStatusAuthorized:
+            complitionHandler(YES);
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
            completionHandler(YES);
```

File: Talkable/TKBLContactsLoader.m [hunk 7]
```diff
@@ -20,87 +21,110 @@
...
+            break;
+        case CNAuthorizationStatusRestricted:
+        case CNAuthorizationStatusDenied:
+        case CNAuthorizationStatusNotDetermined:
+            [store requestAccessForEntityType:CNEntityTypeContacts completionHandler:^(BOOL granted, NSError * _Nullable error) {
+                if (error) {
+                    TKBLLog(@"error while loading contacts - %@", [error localizedDescription]);
+                }
+                
+                complitionHandler(granted);
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
                completionHandler(granted);
```

File: Talkable/TKBLOfferTarget.m [hunk 8]
```diff
@@ -171,7 +171,7 @@ - (void)tkblPutToClipboard:(NSDictionary*)params sender:(id)sender {
 }
 
 - (void)tkblImportContacts:(NSDictionary*)params sender:(id)sender {
-    [[TKBLContactsLoader loader] loadContactsWithComplitionHandler:^(NSArray* contacts) {
+    [[TKBLContactsLoader new] loadContactsWithComplitionHandler:^(NSArray *contacts) {
```
Comment (zhuravel) [inline Talkable/TKBLOfferTarget.m]:
```suggestion
    [[TKBLContactsLoader new] loadContactsWithCompletionHandler:^(NSArray *contacts) {
```

File: Talkable/TKBLContactsLoader.m [hunk 9]
```diff
@@ -20,87 +21,110 @@
...
+            NSString *lastName = contact.familyName;
+            NSMutableArray *phoneNumbers = [NSMutableArray new];
+            NSMutableArray *emailAddresses = [NSMutableArray new];
+
+            if (lastName == nil) {
+                fullName = [NSString stringWithFormat:@"%@", firstName];
+            } else if (firstName == nil){
+                fullName = [NSString stringWithFormat:@"%@", lastName];
+            } else {
+                fullName = [NSString stringWithFormat:@"%@ %@",firstName, lastName];
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
                fullName = [NSString stringWithFormat:@"%@ %@", firstName, lastName];
```

File: Talkable/TKBLContactsLoader.h [hunk 10]
```diff
@@ -16,8 +16,6 @@ FOUNDATION_EXPORT NSString* TKBLContactPhoneNumberKey;
 
 @interface TKBLContactsLoader : NSObject
 
-+ (instancetype)loader;
-
-- (void)loadContactsWithComplitionHandler:(void(^)(NSArray* contactList))complitionHandler;
+- (void)loadContactsWithcompletionHandler:(void(^)(NSArray* contacts))completionHandler;
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.h]:
```suggestion
- (void)loadContactsWithCompletionHandler:(void(^)(NSArray* contacts))completionHandler;
```

File: Talkable/TKBLContactsLoader.m [hunk 11]
```diff
@@ -20,87 +21,110 @@
...
-        TKBLLog(@"error while loading contacts - %@", CFBridgingRelease(error));
-        return;
-    }
-
-    ABAddressBookRequestAccessWithCompletion(addressBook, ^(bool granted, CFErrorRef error) {
-        if (error)
-            TKBLLog(@"error while loading contacts - %@", CFBridgingRelease(error));
-
+- (void)loadContactsWithcompletionHandler:(void(^)(NSArray* contacts))completionHandler {
+    [self requestForAccessWithcompletion:^(BOOL granted) {
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
- (void)loadContactsWithCompletionHandler:(void(^)(NSArray* contacts))completionHandler {
    [self requestForAccessWithCompletion:^(BOOL granted) {
```

File: Talkable/TKBLContactsLoader.m [hunk 12]
```diff
@@ -20,87 +21,110 @@
...
+}
 
-        [contacts addObject:@{
-            TKBLContactFirstNameKey: firstName ? firstName : [NSNull null],
-            TKBLContactLastNameKey: lastName ? lastName : [NSNull null],
-            TKBLContactFullNameKey: fullName ? fullName : [NSNull null],
-            TKBLContactEmailKey: emails,
-            TKBLContactPhoneNumberKey: phoneNumbers
-        }];
+- (void)requestForAccessWithcompletion:(void(^)(BOOL granted))completionHandler {
```
Comment (zhuravel) [inline Talkable/TKBLContactsLoader.m]:
```suggestion
- (void)requestForAccessWithCompletion:(void(^)(BOOL granted))completionHandler {
```

File: Talkable/TKBLOfferTarget.m [hunk 13]
```diff
@@ -171,7 +171,7 @@ - (void)tkblPutToClipboard:(NSDictionary*)params sender:(id)sender {
 }
 
 - (void)tkblImportContacts:(NSDictionary*)params sender:(id)sender {
-    [[TKBLContactsLoader loader] loadContactsWithComplitionHandler:^(NSArray* contacts) {
+    [[TKBLContactsLoader new] loadContactsWithcompletionHandler:^(NSArray *contacts) {
```
Comment (zhuravel) [inline Talkable/TKBLOfferTarget.m]:
```suggestion
    [[TKBLContactsLoader new] loadContactsWithCompletionHandler:^(NSArray *contacts) {
```

File: Talkable/TKBLHelper.m [hunk 14]
```diff
@@ -77,6 +77,16 @@ + (BOOL)isFacebookSharingUsingSocialFrameworkAvailable {
         [[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"fbauth2://"]];;
 }
 
++ (UIViewController*) topMostController {
+    UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
```
Comment (dgtv) [inline Talkable/TKBLHelper.m:81]: In your previous PR I see `topController` could be `null` so looks like you should handle this case:
https://github.com/talkable/talkable-ios-sdk/pull/32/files#diff-875122cdd758244c2ab3c1a9892d8194ea50fbf438ff43aec13a17c50847de8eR45

File: Talkable/TKBLContactsLoader.m [hunk 15]
```diff
@@ -20,87 +21,110 @@
...
-        ABMultiValueRef abEmails = ABRecordCopyValue(abContact, kABPersonEmailProperty);
-        for (CFIndex email_idx = 0; email_idx < ABMultiValueGetCount(abEmails); email_idx++) {
-            NSString* email = CFBridgingRelease(ABMultiValueCopyValueAtIndex(abEmails, email_idx));
-            [emails addObject:email];
+    
+    [store enumerateContactsWithFetchRequest:request error:&error usingBlock:^(CNContact * __nonnull contact, BOOL * __nonnull stop) {
+        if (error) {
+            TKBLLog(@"error while loading contacts - %@", [error localizedDescription]);
+        } else {
+            ;
```
Comment (dgtv) [inline Talkable/TKBLContactsLoader.m]: What is this?

File: Talkable/TKBLContactsLoader.m [hunk 16]
```diff
@@ -20,87 +21,110 @@
...
+            for (CNLabeledValue *label in contact.emailAddresses) {
+                NSString *email = label.value;
+                if ([email length] > 0) {
+                    [emailAddresses addObject:email];
+                }
+            }
+            
+            [contacts addObject:@{TKBLContactFirstNameKey: firstName ? firstName : [NSNull null],
+                                  TKBLContactLastNameKey: lastName ? lastName : [NSNull null],
+                                  TKBLContactFullNameKey: fullName ? fullName  : [NSNull null],
```
Comment (dgtv) [inline Talkable/TKBLContactsLoader.m:96]: Isn't `fullName` always present in current implementation: it equals `[NSString stringWithFormat:@"%@", nil];` if both `firstName` & `lastName ` eq `nil`

**Review by dgtv — CHANGES_REQUESTED** [review]

File: Talkable/TKBLHelper.m (hunk 14 above)
Comment (kostilal) [reply Talkable/TKBLHelper.m:81]: It's a difference between Objective-c and swift. In Objective-c you can send message to nil object. And nothing will happens. But in Swift you can't, so you need check for nil object. Here is more information about Objective - c runtime.

[Sending Messages to nil](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjectiveC/Chapters/ocObjectsClasses.html )

File: Talkable/TKBLContactsLoader.m (hunk 16 above)
Comment (kostilal) [reply Talkable/TKBLContactsLoader.m:96]: You is right, I'l fix it.

**Review by dgtv — APPROVED** [review]

**Review by zhuravel — APPROVED** [review]

**Pull request merged**